File: frequency.ml
(* #load "str.cma";; *)
module StringMap = Map.Make (String)


let file = "./pride_prejudice.txt"
let stop_words_file = "./stop_words.txt"

let read_file (f: string): string = In_channel.with_open_text f In_channel.input_all;;

let split_text (s: string) (sw: string): string list = 
  let r = Str.regexp "[^a-zA-Z]" in
  let words = Str.split r s in 
  let content = read_file sw in
  let stop_words = Str.split r content in
  List.filter (fun x -> String.length x > 1 && not(List.exists (fun y -> y = x) stop_words)) words;;

let count_occ (acc: int StringMap.t) (w: string): int StringMap.t =
  let x = String.lowercase_ascii w in
  if StringMap.mem x acc then
    let old_value = StringMap.find x acc in
    StringMap.add x (old_value+1) acc
  else
    StringMap.add x 1 acc
  ;;

let frequency (f: string) (stop_words_file: string) (n: int): unit =
  let content = read_file f in
  let words = split_text content stop_words_file in
  let count_words = List.fold_left count_occ StringMap.empty words in
  let count_words_list = StringMap.bindings count_words in
  let sorted_count_words_list = List.sort (fun (_, v1) (_, v2) -> v2-v1) count_words_list in
  let n_words = List.fold_left 
    (fun (acc, count) x -> if count < n then (x :: acc, count + 1) else (acc, count)) 
    ([], 0) sorted_count_words_list in
  List.iter (fun k -> Printf.printf "(%s, %d)\n" (fst k) (snd k)) (List.rev (fst n_words));;

if Array.length Sys.argv != 3 then
  Printf.printf ("Il faut deux paramètres, le chemin du fichier txt et le nombre de mots\n")
else
  frequency Sys.argv.(1) stop_words_file (int_of_string Sys.argv.(2))

File: frequency.py

import re
import sys

def path_as_list_of_words(path):
    with open(path, 'r') as f:
        content = f.read()
        words = re.split('[^a-zA-Z]', content)
    return words

def frequency(path_book, path_stop_words, n) : 
    dict = {}
    words = path_as_list_of_words(path_book)
    stop_words = path_as_list_of_words(path_stop_words)
    for w in words: 
        m = w.lower()
        if len(m)>1 and m not in stop_words:
            if m in dict:
                dict[m] += 1
            else:
                dict[m]= 1
    sorted_dict = sorted(dict.items(), key=lambda item:item[1], reverse=True)
    for i in range(n):
        print(sorted_dict[i])

if len(sys.argv) != 3:
    print("Il faut deux paramètres, le chemin du fichier txt et le nombre de mots")
else:
    frequency(sys.argv[1], "stop_words.txt", int(sys.argv[2]))